lib/ansible/modules/cloud/vmware/vmware_guest.py

```diff
@@ -835,53 +846,69 @@ def configure_cpu_and_memory(self, vm_obj, vm_creation=False):
+            for cdrom_pair in zip_longest(cdrom_devices, self.params["cdrom"]):
+                cdrom_device, cdrom_params = cdrom_pair
+                if cdrom_params is not None and not cdrom_params:
+                    # Skip CD-ROM that exists in VM, and has empty (but declared) params.
+                    continue
+
+                if cdrom_params is None:
+                    # CD-ROM is in VM and not params. Delete it
+                    cdrom_spec = vim.vm.device.VirtualDeviceSpec()
+                    cdrom_spec.operation = vim.vm.device.VirtualDeviceSpec.Operation.remove
+                    cdrom_spec.device = cdrom_device
+                    self.configspec.deviceChange.append(cdrom_spec)
                     self.change_detected = True
-                    self.configspec.deviceChange.append(ide_device)
-                elif len(ide_device.device) > 3:
-                    self.module.fail_json(msg="hardware.cdrom specified for a VM or template which already has 4 IDE devices of which none are a cdrom")
-
-                cdrom_spec = self.device_helper.create_cdrom(ide_ctl=ide_device, cdrom_type=self.params["cdrom"]["type"], iso_path=iso_path)
-                if vm_obj and vm_obj.runtime.powerState == vim.VirtualMachinePowerState.poweredOn:
-                    cdrom_spec.device.connectable.connected = (self.params["cdrom"]["type"] != "none")
-
-            elif not self.device_helper.is_equal_cdrom(vm_obj=vm_obj, cdrom_device=cdrom_device, cdrom_type=self.params["cdrom"]["type"], iso_path=iso_path):
-                # Updating an existing CD-ROM
-                if self.params["cdrom"]["type"] in ["client", "none"]:
-                    cdrom_device.backing = vim.vm.device.VirtualCdrom.RemotePassthroughBackingInfo()
-                elif self.params["cdrom"]["type"] == "iso":
-                    cdrom_device.backing = vim.vm.device.VirtualCdrom.IsoBackingInfo(fileName=iso_path)
-                cdrom_device.connectable = vim.vm.device.VirtualDevice.ConnectInfo()
-                cdrom_device.connectable.allowGuestControl = True
-                cdrom_device.connectable.startConnected = (self.params["cdrom"]["type"] != "none")
-                if vm_obj and vm_obj.runtime.powerState == vim.VirtualMachinePowerState.poweredOn:
-                    cdrom_device.connectable.connected = (self.params["cdrom"]["type"] != "none")
-
-                cdrom_spec = vim.vm.device.VirtualDeviceSpec()
-                cdrom_spec.operation = vim.vm.device.VirtualDeviceSpec.Operation.edit
-                cdrom_spec.device = cdrom_device
-
-            if cdrom_spec:
-                self.change_detected = True
-                self.configspec.deviceChange.append(cdrom_spec)
+                    continue
+
+                if "type" not in cdrom_params or cdrom_params["type"] not in ["none", "client", "iso"]:
+                    self.module.fail_json(msg="cdrom.type is mandatory")
```

Comment:
Maybe suggesting existing/available options in error message will be help to end user.